controllers/dashboard_controller.go

```diff
@@ -0,0 +1,329 @@
+			continue
+		}
+
+		// first reconcile the plugins
+		// append the requested dashboards to a configmap from where the
+		// grafana reconciler will pick them upi
+		err = r.reconcilePlugins(ctx, &grafana, dashboard)
+		if err != nil {
+			controllerLog.Error(err, "error reconciling plugins", "dashboard", dashboard.Name, "grafana", grafana.Name)
+		}
+
+		// then import the dashboard into the matching grafana instances
+		err = r.reconcileDashboardInInstance(ctx, &grafana, dashboard)
+		if err != nil {
+			controllerLog.Error(err, "error reconciling dashboard", "dashboard", dashboard.Name, "grafana", grafana.Name)
+		}
+	}
+
+	return ctrl.Result{RequeueAfter: RequeueDelaySuccess}, nil
+}
+
+func (r *GrafanaDashboardReconciler) reconcileDashboardInInstance(ctx context.Context, grafana *v1beta1.Grafana, cr *v1beta1.GrafanaDashboard) error {
+	if strings.TrimSpace(cr.Spec.Json) == "" {
+		return nil
+	}
+
+	var dashboardFromJson map[string]interface{}
+
+	err := json.Unmarshal([]byte(cr.Spec.Json), &dashboardFromJson)
+	if err != nil {
+		return err
+	}
+
+	grafanaClient, err := client2.NewGrafanaClient(ctx, r.Client, grafana)
+	if err != nil {
+		r.Log.Error(err, "error establishing new api client for instance", "grafana", grafana.Name, "namespace", grafana.Namespace, "dashboard", cr.Name)
+		return err
+	}
+
+	dashboards, err := grafanaClient.Dashboards()
+	if err != nil {
+		r.Log.Error(err, "could not get dashboards")
+		return err
+	}
+	logger.Log.Info("found dashboards", "dashboards", dashboards)
+
+	status := cr.Status.DeepCopy()
+
+	if status.UIDforInstance == nil {
+		status.UIDforInstance = make(map[string]string)
+	}
+
+	getDashUIDFromStatus := func(namespacedName string) string {
+		if status != nil {
+			for grafanaInstance, dashboardUID := range status.UIDforInstance {
+				if grafanaInstance == namespacedName {
+					return dashboardUID
+				}
+			}
+		}
+		return ""
+	}
+
+	grafanaNamespacedName := fmt.Sprintf("%s%s", grafana.Namespace, grafana.Name)
+	dashUIDfromStatus := getDashUIDFromStatus(grafanaNamespacedName)
+
+	if cr.DeletionTimestamp != nil {
+		r.Log.Info("found a deletion timestamp for dashboard", "dashboard", cr.ObjectMeta.Name, "instance", grafanaNamespacedName)
+		if err := grafanaClient.DeleteDashboardByUID(dashUIDfromStatus); strings.Contains(fmt.Sprintf("%v", err), "404") {
+			r.Log.Info("the dashboard couldn't be found on the instance, nothing to do, removing finalizer")
+			controllerutil.RemoveFinalizer(cr, dashboardFinalizer)
+			err = r.Client.Update(ctx, cr)
+			if err != nil {
+				return err
+			}
+			return nil
+		}
+		if err != nil {
+			r.Log.Error(err, "could not delete dashboard from instance", "instance", grafanaNamespacedName)
+			return err
+		} else {
+			delete(status.UIDforInstance, grafanaNamespacedName)
+
+			if err := r.reconcileDashboardStatus(cr, status); err != nil {
+				r.Log.Error(err, "error while updating dashboard status")
+				return err
+			}
+			if status == nil {
```

Comment:
when would the status be nil? When the last dashboard is removed, the map is empty, but not nil. Is a `len(status) == 0` missing here?